src/jrd/cch.cpp

```diff
@@ -4316,60 +4232,74 @@ static ULONG memory_init(thread_db* tdbb, BufferControl* bcb, SLONG number)
 					// but it's not virtually contiguous. Let's find out by
 					// cutting the size in half to see if the buffers can be
 					// scattered over the remaining virtual address space.
-					memory_size >>= 1;
-					if (memory_size < MIN_BUFFER_SEGMENT)
-					{
-						// Diminishing returns
-						return buffers;
-					}
+					to_alloc >>= 1;
 				}
 			}
-
 			bcb->bcb_memory.push(memory);
-			memory_end = memory + memory_size;
+
+			tail = (BufferDesc*) FB_ALIGN(memory, alignof(BufferDesc));
+
+			BufferControl::BDBBlock blk;
+			blk.m_bdbs = tail;
+			blk.m_count = to_alloc;
+			bcb->bcb_bdbBlocks.push(blk);
```

Comment:
I suspect some memory overhead is possible here, given that our Array always doubles its storage size when expanding. Maybe bcb_bdbBlocks should be pre-allocated with the known maximum size, e.g. BufferControl::create(dbb, number) where number is passed to BufferControl::ctor and then into bcb_bdbBlocks(p, number)?